File: 098_ValidateBinarySearchTree_2.cpp
// Topic: Validate Binary Search Tree
// Description: Determine if a binary tree is a valid BST.
// Approach: Iterative inorder traversal to ensure values are strictly increasing.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    long long val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(long long x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTreeLevel(const vector<long long>& vals)
{
    if (vals.empty())
        return nullptr;
    int n = vals.size();
    if (vals[0] == LLONG_MIN)
        return nullptr;
    TreeNode* root = new TreeNode(vals[0]);
    queue<TreeNode*> q;
    q.push(root);
    int i = 1;
    while (!q.empty() && i < n)
    {
        TreeNode* node = q.front();
        q.pop();
        if (i < n)
        {
            if (vals[i] != LLONG_MIN)
            {
                node->left = new TreeNode(vals[i]);
                q.push(node->left);
            }
            ++i;
        }
        if (i < n)
        {
            if (vals[i] != LLONG_MIN)
            {
                node->right = new TreeNode(vals[i]);
                q.push(node->right);
            }
            ++i;
        }
    }
    return root;
}

bool isValidBST(TreeNode* root)
{
    stack<TreeNode*> st;
    TreeNode* cur = root;
    long long prev = LLONG_MIN;
    bool first = true;
    while (cur || !st.empty())
    {
        while (cur)
        {
            st.push(cur);
            cur = cur->left;
        }
        cur = st.top();
        st.pop();
        if (!first && cur->val <= prev)
            return false;
        prev = cur->val;
        first = false;
        cur = cur->right;
    }
    return true;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<long long>> tests = {{2, 1, 3},                              // valid BST
                                       {5, 1, 4, LLONG_MIN, LLONG_MIN, 3, 6},  // invalid BST
                                       {}};

    for (int i = 0; i < (int)tests.size(); ++i)
    {
        TreeNode* root = buildTreeLevel(tests[i]);
        cout << "Test " << (i + 1) << ": " << (isValidBST(root) ? 1 : 0) << '\n';
    }
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/

File: 098_ValidateBinarySearchTree_2.py
# Topic: Validate Binary Search Tree
# Description: Determine if a binary tree is a valid BST.
# Approach: Iterative inorder traversal to ensure values are strictly increasing.

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        stack = []
        cur = root
        prev = float('-inf')
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            if cur.val <= prev:
                return False
            prev = cur.val
            cur = cur.right
        return True


if __name__ == "__main__":
    # Test 1: Valid BST
    root = TreeNode(2, TreeNode(1), TreeNode(3))
    print(f"Test 1: {Solution().isValidBST(root)}")

    # Test 2: Invalid BST
    root = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    print(f"Test 2: {Solution().isValidBST(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""
